ansible/roles/monitoring/templates/ecs-discovery.py

```diff
@@ -0,0 +1,275 @@
+
+    def _set_ec2_private_ips(self, services: List[ECSService]):
+        """
+        Set up the private IP for the given list of services
+        """
+
+        ec2_client = boto3.client(
+            "ec2",
+            aws_access_key_id=self.access_key,
+            aws_secret_access_key=self.secret_key,
+            region_name=self.region,
+        )
+        instance_ids = [service.ec2_instance_id for service in services]
+        instance_description = ec2_client.describe_instances(InstanceIds=instance_ids)
+
+        instances = {}
+        for reservation in instance_description["Reservations"]:
+            for instance in reservation["Instances"]:
+                instance_id = instance["InstanceId"]
+                instances[instance_id] = instance
+
+        for service in services:
+            private_ip = instances[service.ec2_instance_id]["PrivateIpAddress"]
+            service.private_ip = private_ip
+
+def to_prom_json(services : List[ECSService]) -> List[Dict[str, Any]]:
+    """
+    Convert a list of service objects into a prometheus-compatible list of dict
+    """
+    services_json = []
+
+    for service in services:
+        services_json.append({
+            "targets" : [
+                f"{service.private_ip}:{service.port}"
+            ],
+            "labels": {
+                "job" : service.container_name,
+                "instance" : service.ec2_instance_id,
+                "task" : service.task_arn,
+                "date_discovered" : service.date_discovered.isoformat()
+            }
+        })
+
+    return services_json
+
+
+def main(args : argparse.Namespace):
+
+    logging.basicConfig(level=logging.INFO)
+
+    # Collect arguments
+    secret_key = args.secret_key or os.environ.get("AWS_SECRET_KEY")
+    access_key = args.access_key or os.environ.get("AWS_ACCESS_KEY_ID")
+    region = args.region or os.environ.get("AWS_REGION")
+
+    # Check that all arguments are passed 
+    mandatory_args = [('secret key', secret_key), ('access key', access_key), ('region', region)]
+    for (arg_name, arg_val) in mandatory_args:
+        if arg_val is None:
+            logging.error(f"Missing argument: {arg_name}. You can specify it by command line arguments or environment variables, see --help")
+            exit(1)
+    
+    # If no show and no output file, do nothing 
+    if args.show and args.output_file == "None":
+        return
+    
+    discovery = ECSDiscovery(region, secret_key, access_key) # type: ignore
+    services = discovery.list_services()
+
+    if args.show:
+        for service in services:
+            logging.info(f"[Cluster {service.cluster}] ({service.container_name}) {service.private_ip}:{service.port}")
+
+    # Save file to disk
+    if args.output_file != "None":
+        services_json = to_prom_json(services)
+        path = Path(args.output_file)
+        with path.open("w") as f:
+            json.dump(services_json, f)
+            
+
+if __name__ == "__main__":
```

Comment:
This is great, though it's starting to be quite a long script. I wonder if there is something we can do to do some testing of this to make sure it's working as intended. Maybe with some of the dev credentials that we are inside of the CI here.

Reply:
I agree that it would be nice to have some testing for this, but how could we do it? since the bulk of the work in this script is pulling data from AWS

Or maybe we start with something simple like ensuring that it can list whatever is in AWS without crashing?

Reply:
Maybe we can use the dev aws key which we use for the CI and ensure it's able to write an appropriate json file. We can do this in future work, but let's note it in an issue